server/src/main/java/org/spine3/server/entity/storage/EntityQueryMatcher.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.entity.storage;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Multimap;
+import com.google.protobuf.Any;
+import org.spine3.annotations.Internal;
+import org.spine3.client.EntityId;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Collections2.transform;
+import static com.google.common.collect.Sets.newHashSet;
+
+/**
+ * A {@link Predicate} on the {@link EntityRecordWithColumns} matching it upon the given
+ * {@link EntityQuery}.
+ *
+ * <p>See {@link EntityQuery} for the matching contract
+ *
+ * @author Dmytro Dashenkov
+ */
+@Internal
+public final class EntityQueryMatcher implements Predicate<EntityRecordWithColumns> {
```

Comment:
I hope to avoid using this predicate, as well as any other `Predicate`, in the `GAE` Storage implementation. The point is to perform the queries using the database "native" means.

And if we would consider filtering the records in the memory in other Storage impls., we would probably have a single behavior for all of them.

Thinking of the problems which may appear while trying to query the `GAE` Datastore, I assume that we _may_ want to use the in-memory filtering, but then we'll try to optimize it as mush as possible and there would be no sense in extending this particular `Predicate` implementation.